core/src/main/java/org/apache/iceberg/TableMetadataParser.java

```diff
@@ -252,7 +253,7 @@ static TableMetadata fromJson(TableOperations ops, InputFile file, JsonNode node
     List<Snapshot> snapshots = Lists.newArrayListWithExpectedSize(snapshotArray.size());
     Iterator<JsonNode> iterator = snapshotArray.elements();
     while (iterator.hasNext()) {
-      snapshots.add(SnapshotParser.fromJson(ops, iterator.next()));
+      snapshots.add(SnapshotParser.fromJson(io, TableMetadata.indexSpecs(specs), iterator.next()));
```

Comment:
If we remove the spec map from Snapshot, then this doesn't need to pass in the specs and `indexSpecs` doesn't need to be public.